src/providers/pronouns/Pronouns.cpp

```diff
@@ -0,0 +1,64 @@
+#include "providers/pronouns/Pronouns.hpp"
+
+#include "Application.hpp"
+#include "common/QLogging.hpp"
+#include "providers/pronouns/alejo/PronounsAlejoApi.hpp"
+#include "providers/pronouns/UserPronouns.hpp"
+
+#include <mutex>
+#include <string>
+#include <unordered_map>
+
+namespace chatterino::pronouns {
+
+void Pronouns::fetch(const QString &username,
+                     const std::function<void(UserPronouns)> &callbackSuccess,
+                     const std::function<void()> &callbackFail)
+{
+    // Only fetch pronouns if we haven't fetched before.
+    {
+        std::shared_lock<std::shared_mutex> lock(this->mutex);
+
+        auto iter = this->saved.find(username);
+        if (iter != this->saved.end())
+        {
+            callbackSuccess(iter->second);
+            return;
+        }
+    }  // unlock mutex
+
+    qCDebug(chatterinoPronouns)
+        << "Fetching pronouns from alejo.io for " << username;
+
+    alejoApi.fetch(username, [this, callbackSuccess, callbackFail,
```

Comment:
Oh, I forgot this: As part of the "cleanly exit Chatterino" effort: This shouldn't capture `this`, but a `weak_ptr` to `this` (obtained through [`weak_from_this`](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this/weak_from_this); and the class has to inherit from [`std::enable_shared_from_this`](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this)). 

Similarly, for `AlejoApi`, which shouldn't capture `this` in network requests. It could store a `std::weak_ptr<void>` to its parent (`Pronouns`) on construction and use that to ensure it's still alive when receiving a response for network requests.

But that isn't something you have to address - I suspect pajlada will take care of that before merging.